service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/CloneReferencedResourceStep.java

```diff
@@ -0,0 +1,133 @@
+            .setSourceResourceId(resource.getResourceId())
+            .setName(resource.getName())
+            .setDescription(resource.getDescription());
+    if (CloningInstructions.COPY_REFERENCE == resource.getCloningInstructions()) {
+      FlightUtils.validateRequiredEntries(
+          context.getInputParameters(),
+          ControlledResourceKeys.DESTINATION_WORKSPACE_ID,
+          JobMapKeys.AUTH_USER_INFO.getKeyName());
+      var destinationWorkspaceId =
+          context
+              .getInputParameters()
+              .get(ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+
+      ReferencedResource destinationResource =
+          resource
+              .buildReferencedClone(
+                  destinationWorkspaceId,
+                  destinationResourceId,
+                  destinationFolderId,
+                  resource.getName(),
+                  resource.getDescription())
+              .castToReferencedResource();
+
+      cloneDetails
+          .setResourceType(destinationResource.getResourceType())
+          .setStewardshipType(destinationResource.getStewardshipType())
+          .setCloningInstructions(destinationResource.getCloningInstructions())
+          .setDestinationResourceId(destinationResource.getResourceId());
+
+      // put the destination resource in the map, because it's not communicated
+      // from the flight as the response (and we need the workspace ID)
+      context
+          .getWorkingMap()
+          .put(ControlledResourceKeys.DESTINATION_REFERENCED_RESOURCE, destinationResource);
+
+      try {
+        resourceDao.createReferencedResource(destinationResource);
+        cloneDetails.setResult(WsmCloneResourceResult.SUCCEEDED);
+      } catch (Exception e) {
+        cloneDetails.setResult(WsmCloneResourceResult.FAILED).setErrorMessage(e.getMessage());
+      }
+    } else {
+      cloneDetails
+          .setResult(WsmCloneResourceResult.SKIPPED)
+          .setCloningInstructions(resource.getCloningInstructions())
+          .setDestinationResourceId(null)
+          .setErrorMessage(null);
+    }
+    // add to the map
+    addCloneDetailsToWorkingMap(context, cloneDetails);
+    return StepResult.getStepResultSuccess();
+  }
+
+  private void addCloneDetailsToWorkingMap(
+      FlightContext context, WsmResourceCloneDetails cloneDetails) {
+    var resourceIdToResult =
+        Optional.ofNullable(
+                context
+                    .getWorkingMap()
+                    .get(
+                        ControlledResourceKeys.RESOURCE_ID_TO_CLONE_RESULT,
+                        new TypeReference<Map<UUID, WsmResourceCloneDetails>>() {}))
+            .orElseGet(HashMap::new);
+    resourceIdToResult.put(resource.getResourceId(), cloneDetails);
+    context
+        .getWorkingMap()
+        .put(ControlledResourceKeys.RESOURCE_ID_TO_CLONE_RESULT, resourceIdToResult);
+  }
+
+  // No need to undo here as entire workspace will be destroyed by earlier undo
```

Comment:
the code makes sense but does make it seem like the comment is out of date

Reply:
removed the comment. I think it's bad practice to not undo this just because a previous step (in another flight even!) that will delete the workspace. Step should be modular. If this flight isn't get launched as subflight of cloneGcpWorkspace, we will not delete this resource..